File: BST.py
class Node:
    def __init__(self, key):
        self.left = None
        self.right = None
        self.val = key 

class BinaraySearchTree:
    def __init__(self):
        self.root = None

    def insert(self, key):
        if self.root is None:
            self.root = Node(key)
        else:
            self._insert_recursive(self.root, key)
    
    def _insert_recursive(self, root, key):
        if key < root.val:
            if root.left is None:
                root.left = Node(key)
            else:
                self._insert_recursive(root.left, key)
        
        else:
            if root.right is None:
                root.right = Node(key)
            else:
                self._insert_recursive(root.right, key)

    def search(self, key):
        return self._search_recursive(self.root, key)
    
    def _search_recursive(self, root, key):
        if root is None or root.val == key:
            return  root
        if key < root.val:
            return self._search_recursive(root.left, key)
        return self._search_recursive(root.right, key)
    
    def inorder(self):
        self._inorder(self.root)
    
    def _inorder(self, root):
        if root:
            self._inorder(root.left)
            print(root.val, end='')
            self._inorder(root.right)

# Example Usage:

bst = BinaraySearchTree()
bst.insert(50)
bst.insert(30)
bst.insert(20)
bst.insert(40)
bst.insert(70)
bst.insert(60)
bst.insert(80)

print("Inorder Traversal of the BST: ")
bst._inorder()

File: BST.cpp
#include<iostream>
using namespace std;

//  Node structure for the BST

struct Node{
    int key;
    Node* left;
    Node *right;

    Node(int k): key(k), left(nullptr), right(nullptr){}
};

// BST Class

class BinarySearrchTree{

public:
    BinarySearrchTree(): root(nullptr){}

    void insert(int key){
        root = insertRecursive(root,key);
    }

    bool search(int key){
        return searchRecursive(root,key) != nullptr;
    }

    void inorder(){
        inorderRecursive(root);
        cout << endl;
    }

private:
    Node* root;

    Node* insertRecursive(Node* node, int key){
        if(node == nullptr){
            return new Node(key);
        }

        if(key < node->key){
            node->left = insertRecursive(node->left, key);
        }
        else{
            node->right = insertRecursive(node->right,key);
        }
        
        
        return node;
    }

    Node* searchRecursive(Node* node,int key){
        if(node == nullptr || node->key == key){
            return node;
        }

        if(key < node->key){
            return searchRecursive(node->left,key);
        }

        if(key > node->key){
            return searchRecursive(node->right,key);
        }
    }

    void inorderRecursive(Node* node){
        if(node != nullptr){
            inorderRecursive(node->left);
            cout << node->key << " ";
            inorderRecursive(node->right);
        }
    }

};

int main(){
    BinarySearrchTree bst;
    bst.insert(50);
    bst.insert(30);
    bst.insert(20);
    bst.insert(40);
    bst.insert(70);
    bst.insert(60);
    bst.insert(80);


    cout << "Inorder Treaversal of the BST: ";
    bst.inorder(); // OUTPUT: 20 30 40 50 60 70 80

    cout << "Search for 40 : " << (bst.search(40) ? "Found" : "Not Found ") << endl;
    cout << "Search for 25: " << (bst.search(25) ? "Found" : "Not Found ") << endl;

    return 0;

}
